### PR TITLE
Refactor search feature

### LINKED ISSUE #81: Dash `-` crashes the search
When I type either `-` (without anything else), or `--` (after something else for example `test--`). The page crashes showing this screen:
image

### PR DESCRIPTION
**Pull request by damianstasik**
TODO:
- [x] Use generated types once they are ready
- [x] Make sure correct titles are used (backend changes required)
- [x] Finish UI for all search type results
- [x] Integrate the search feature on home page
- [x] Hide the search bar in the header while being on the home page

Fixes #81, #59 and #48

### CODE REVIEW

File: frontend/src/components/Search/index.tsx [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+              ))}
+            </div>
+          ))}
+        </ComboboxOptions>
+      </div>
+    </Combobox>
+  );
+}
+
+Search.defaultProps = {
```
Comment (damianstasik) [inline frontend/src/components/Search/index.tsx:259]: `defaultProps` are getting removed from React 19. I'd suggest providing these as default values on the 115th line.

File: frontend/src/q.ts [hunk 2]
```diff
@@ -1,13 +1,29 @@
...
+            // and utilizing the signal to cancel the request if the query changes
+            // we get a debounce effect
+            await (() => new Promise((resolve) => setTimeout(resolve, 100)))();
+            if (signal.aborted) {
+              return;
+            }
 
-      return lunr.Index.load(data);
-    },
+            const response = await api(
```
Comment (damianstasik) [inline frontend/src/q.ts:19]: We only need to await `response.json`, and not even that since we can return it and let React Query handle it.

Comment (Yantrio) [reply frontend/src/q.ts:19]: Okay

File: frontend/src/q.ts [hunk 3]
```diff
@@ -1,13 +1,24 @@
...
+              return;
+            }
 
-      return lunr.Index.load(data);
-    },
+            return await api(`search?q=${encodeURIComponent(query)}`, {
+              signal,
+            }).json<definitions["SearchResultItem"][]>();
+          }
+        : skipToken,
```
Comment (RLRabinowitz) [inline frontend/src/q.ts:23]: Shouldn't this approach send a request for every character you type, even if you do it faster than the debounce (100ms)?

It should probably be better to debounce the query key - https://stackoverflow.com/questions/76030996/how-to-use-debounce-with-usequery-in-react-query

**Review by RLRabinowitz — COMMENTED** [review]
Looks good overall

File: frontend/src/q.ts (hunk 3 above)
Comment (damianstasik) [reply frontend/src/q.ts:23]: If we want to debounce it I'd suggest moving this out of the query and replacing the `useDeferredValue` (in the Search component) with a simple hook that would do the debouncing. We could either use something from npm or do something simple like:
```ts
function useDebouncedValue(value: string, delay: number) {
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);

    return () => {
      clearTimeout(handler);
    };
  }, [value, delay]);

  return debouncedValue;
}
```

Comment (Yantrio) [reply frontend/src/q.ts:23]: I was under the assumption that the tanstack query signal that gets sent down handles the cancellation. By me checking `if (signal.aborted)` im checking if tanstack aborted the query so i can move on. 

But I could be wrong in these assumptions.

Comment (damianstasik) [reply frontend/src/q.ts:23]: In theory `fetch` shouldn't even start if a signal is already aborted.

Comment (RLRabinowitz) [reply frontend/src/q.ts:23]: Thank you, wasn't aware of this

**Review by RLRabinowitz — APPROVED** [review]

**Pull request merged**